packages/web-sdk/src/instrumentations/session/instrumentation.ts

```diff
@@ -103,17 +91,42 @@ export class SessionInstrumentation extends BaseInstrumentation {
     const sessionTracking = this.config.sessionTracking;
 
     if (sessionTracking?.enabled) {
+      const SessionManager = this.config.sessionTracking?.persistent
+        ? PersistentSessionsManager
+        : VolatileSessionsManager;
+
       const initialSessionMeta = this.createInitialSessionMeta(sessionTracking);
+
+      SessionManager.storeUserSession(
+        createUserSessionObject({
+          sessionId: initialSessionMeta.id,
+          isSampled: Boolean(initialSessionMeta.attributes!['isSampled']),
+        })
+      );
+
+      this.notifiedSession = initialSessionMeta;
       this.api.setSession(initialSessionMeta);
 
-      const sessionManager = this.getSessionManagerInstanceByConfiguredStrategy(this.metas.value.session?.id);
+      const { updateSession } = new SessionManager();
+
+      this.transports?.addBeforeSendHooks((item) => {
+        updateSession();
+
+        const attributes = item.meta.session?.attributes;
+
+        if (attributes && Boolean(attributes?.['isSampled'])) {
+          const { isSampled: _, ...restAttributes } = attributes;
+
+          item.meta.session = {
+            ...item.meta.session,
+            attributes: restAttributes,
+          };
 
-      if (sessionManager != null) {
-        this.transports?.addBeforeSendHooks(...this.transports.getBeforeSendHooks(), (item: any) => {
-          sessionManager?.updateSession();
           return item;
-        });
-      }
+        }
+
+        return null;
```

Comment:
I don't understand how this can work with batching

if batching is enabled, `execute` both enqueues it _and_ calls `instantSend`: https://github.com/grafana/faro-web-sdk/blob/9dca6e8e99e32cade68a700059cae515134902e5/packages/core/src/transports/initialize.ts#L147-L166

`instantSend` then applies this filter: https://github.com/grafana/faro-web-sdk/blob/9dca6e8e99e32cade68a700059cae515134902e5/packages/core/src/transports/initialize.ts#L121

It will then remove the `isSampled` attr (return on line 125)

It then loops over all transports, and only sends if the transport returns `false` for `isBatched()`

https://github.com/grafana/faro-web-sdk/blob/9dca6e8e99e32cade68a700059cae515134902e5/packages/core/src/transports/initialize.ts#L127-L134

We only have a batched transport, so nothing happens.

---

Then we wait for the batching interval to trigger: https://github.com/grafana/faro-web-sdk/blob/9dca6e8e99e32cade68a700059cae515134902e5/packages/core/src/transports/batchExecutor.ts#L54

This invokes `sendFn`, which is `batchedSend`: https://github.com/grafana/faro-web-sdk/blob/9dca6e8e99e32cade68a700059cae515134902e5/packages/core/src/transports/initialize.ts#L105-L118

It then does the _same_ `applyBeforeSendHooks` on the item that `instantSend` has already filtered, so it hits this `return null`, meaning it doesn't send the item to the backend.

However, `instantSend` never actually sent it to the backend, so the item is never actually sampled.

---

I think what needs to happen is that `item.meta.session.attributes.isSampled` is never removed until the item is actually sent to the backend, not during the _check_ for whether or not it's sampled.

Reply:
maybe something like this: https://github.com/SimenB/faro-web-sdk/commit/39f62ad5dee44a045a81ed79f1b5fb99be002e46

Reply:
Hey Simen thanks a lot again and you are right, very good catch! 🙏 

I'll fix this and will change the cleanup logic.
The important thing I would like to avoid is that a specific transport need to take care about cleaning up the meta. 

Cheers,
Marco

Reply:
But yes, letting batchExecuter be responsible to clone will prevent similar issues to arise.

Reply:
> maybe something like this: [SimenB@39f62ad](https://github.com/SimenB/faro-web-sdk/commit/39f62ad5dee44a045a81ed79f1b5fb99be002e46)

Oh I am so sorry totally missed that.
Give me a few minutes to look at it.

And thanks a lot again, the contributions, reviews and input is great and super helpful!

Reply:
Found another bug:

[`if (attributes && Boolean(attributes?.['isSampled']))`](https://github.com/grafana/faro-web-sdk/blob/9c729be7e4c315d9ee75ec0ae5678c8f4a2f355a/packages/web-sdk/src/instrumentations/session/instrumentation.ts#L117-L118) 
won`t work because attributes are always strings.